lib/atscppapi/examples/websocket/WebSocket.cc

```diff
@@ -0,0 +1,246 @@
+{
+    std::string frame(1, char(code));
+    size_t len = data.length();
+    if (len <= 125) {
+        frame += char(len);
+    } else {
+        int len_len;
+        if (len <= 65535) {
+            frame += char(126);
+            len_len = 2;
+        } else {
+            frame += char(127);
+            len_len = 8;
+        }
+        while (--len_len >= 0) {
+            frame += char((len >> (8*len_len)) & 0xFF);
+        }
+    }
+    produce(frame + data);
+}
+
+void WebSocket::consume(const std::string &data, InterceptPlugin::RequestDataType type)
+{
+    if (isWebsocket()) {
+        if (!ws_handshake_done_) {
+            std::string digest = ws_digest(ws_key_);
+
+            printf("WS key digest: %s %s\n", ws_key_.c_str(), digest.c_str());
+
+            std::string headers;
+            headers +=
+                "HTTP/1.1 101 Switching Protocols\r\n"
+                "Upgrade: websocket\r\n"
+                "Connection: Upgrade\r\n"
+                "Sec-WebSocket-Accept: " + digest + "\r\n\r\n";
+            produce(headers);
+            ws_handshake_done_ = true;
+        }
+    }
+
+    // auto tid = std::this_thread::get_id();
+    if (type == InterceptPlugin::REQUEST_HEADER) {
+        headers_ += data;
+    } else {
+        // cout << tid << ": Read request body" << endl << data << endl;
+        if (isWebsocket()) {
+
+	    // When we've read as much as we can for any incoming
+	    // data chunk we remove the head of ws_buf_ and reset pos_
+	    // to 0.
+	    //
+	    // If we can't read a complete length + masks, we leave
+	    // pos_ unchanged and just append what we got to ws_buf_.
+	    //
+	    // When we get the length, we update pos_ and length.
+	    // While length is non-zero we wait for ws_buf_ to have
+	    // msg_len_ bytes beyond pos_.
+
+            ws_buf_ += data;
+            while (true) {
+                if (!msg_len_) {
+                    size_t avail = ws_buf_.size() - pos_;
+
+                    // Read the msg_length if we have enough data.
+                    if (avail < 6) break; // 2 + 4 mask
+                    ctrl_ = ws_buf_[pos_] & 0xF;
+                    size_t msg_len = ws_buf_[pos_ + 1] & 0x7F;
+                    if (msg_len == 0x7E) {
+                        if (avail < 8) { // 2 + 2 + length bytes + 4 mask.
+                            break;
+                        }
+                        msg_len = 0;
+                        pos_ += 2;
+                        for (int i = 0; i < 2; ++i, ++pos_) {
+                            msg_len <<= 8;
+                            msg_len += (unsigned char)(ws_buf_[pos_]);
+                        }
+                    } else if (msg_len == 0x7F) {
+                        if (avail < 14) { // 2 + 8 length bytes + 4 mask.
+                            break;
+                        }
+                        msg_len = 0;
+                        pos_ += 2;
+                        for (int i = 0; i < 8; ++i, ++pos_) {
```

Comment:
Similarly you need to convert the 64bit quantity using network byte order conversions:

be64toh((\*((uint64_t\* )(ws_buf_ + pos_))));